### PR TITLE
feat(watermark): Clean state in DynamicFilter by watermark in right side

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Summarize your change (**mandatory**)
Clean state in `DynamicFilter` by watermark **in right side**
- How does this PR work? Need a brief introduction for the changed logic (optional)
Do range delete by cached watermark among corresponding vnodes on barrier
- Describe clearly one logical change and avoid lazy messages (optional)
Buffer and state table will both be cleaned
`BarrierAlign` is modified
- Describe any limitations of the current code (optional)
If comparator is `LessThan`, it will not be cleaned

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#6472

### CODE REVIEW

File: src/stream/src/common/table/state_table.rs [hunk 1]
```diff
@@ -691,7 +720,27 @@ impl<S: StateStore> StateTable<S> {
                 }
             }
         }
+        if let Some(range_end_suffix) = range_end_suffix {
```
Comment (Little-Wallace) [inline src/stream/src/common/table/state_table.rs:723]: Does it means that we must delete-ranges for every epoch ?

File: src/stream/src/common/table/state_table.rs [hunk 2]
```diff
@@ -630,18 +637,24 @@ impl<S: StateStore> StateTable<S> {
         self.epoch = Some(new_epoch);
     }
 
-    pub async fn commit(&mut self, new_epoch: EpochPair) -> StreamExecutorResult<()> {
+    pub async fn commit(
+        &mut self,
+        new_epoch: EpochPair,
+        watermark: Option<&ScalarImpl>,
```
Comment (Little-Wallace) [inline src/stream/src/common/table/state_table.rs:643]: I suggest an independent interface to pass this parameter and update `last_watermark`.

**Comment by jon-chuang** [comment]
Is there a way to add some unit tests based on watermark and check the behaviour? (maybe we can read directly from storage to see if the watermark is deleted, and also check the range cache's range is updated.

File: src/stream/src/common/table/state_table.rs [hunk 3]
```diff
@@ -660,12 +673,28 @@ impl<S: StateStore> StateTable<S> {
         &mut self,
         buffer: BTreeMap<Vec<u8>, RowOp>,
         epoch: u64,
+        watermark: Option<&ScalarImpl>,
```
Comment (jon-chuang) [inline src/stream/src/common/table/state_table.rs:676]: This is a common change as seen in https://github.com/risingwavelabs/risingwave/pull/6330. Can we make a PR just focused on the state table change and add some unit tests for the changes in `test_streaming_table.rs`?

Comment (jon-chuang) [reply src/stream/src/common/table/state_table.rs:676]: I think it'll make it easier to review the changes in one place.

Comment (soundOfDestiny) [reply src/stream/src/common/table/state_table.rs:676]: > This is a common change as seen in #6330. Can we make a PR just focused on the state table change and add some unit tests for the changes in `test_streaming_table.rs`?

just created an independent PR #6495 for this. cc @Little-Wallace

File: src/stream/src/common/table/state_table.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/stream/src/common/table/state_table.rs:737]: > Does it means that we must delete-ranges for every epoch ?

fixed

**Comment by soundOfDestiny** [comment]
> Is there a way to add some unit tests based on watermark and check the behaviour? (maybe we can read directly from storage to see if the state below watermark is deleted, and also check the range cache's range is updated).

I have just added a unit test.

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 4]
```diff
@@ -415,8 +449,44 @@ fn range_contains_lower_upper(
...
+            .await
+            .unwrap();
+        assert_eq!(
+            range_cache.range,
+            Some((Bound::Unbounded, Bound::Unbounded))
+        );
+        range_cache.shrink(&ScalarImpl::Int64(5));
+        assert_eq!(
+            range_cache.range,
+            Some((Bound::Included(ScalarImpl::Int64(5)), Bound::Unbounded))
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:486]: Can we be slightly more comprehensive covering the behaviours?

Comment (soundOfDestiny) [reply src/stream/src/executor/managed_state/dynamic_filter.rs:520]: > Can we be slightly more comprehensive covering the behaviours?

fixed

Comment (jon-chuang) [reply src/stream/src/executor/managed_state/dynamic_filter.rs:520]: Thanks, much better.

File: src/stream/src/common/table/state_table.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/stream/src/common/table/state_table.rs]: > I suggest an independent interface to pass this parameter and update `last_watermark`.

fixed

**Comment by soundOfDestiny** [comment]
> Can this be mergerd now?

I have no objection. plz ask others which is active in this thread

File: src/stream/src/executor/dynamic_filter.rs [hunk 5]
```diff
@@ -300,6 +300,9 @@ impl<S: StateStore> DynamicFilterExecutor<S> {
             left_to_output,
         )?;
 
+        let is_watermark_direct_accord = !matches!(self.comparator, LessThan | LessThanOrEqual);
```
Comment (jon-chuang) [inline src/stream/src/executor/dynamic_filter.rs:303]: can we use an easier to understand variable name.

e.g. `can_clean_state`

**Review by jon-chuang — APPROVED** [review]
Rest LGTM

**Pull request merged**